File: day2/part1.ts
export {};

const inFile = Bun.file("input.txt");

inFile.text().then(text => {
    const lines = text.split('\r\n');

    var result = 0;

    lines.forEach(line => {
        const numbers = line.split(' ').map(el => parseInt(el));
        const increasing = (numbers[1] - numbers[0]) > 0;
        var safe = true;

        for(var i = 1; i < numbers.length; ++i) {
            const diff = numbers[i] - numbers[i-1];
            const absDiff = Math.abs(diff);

            if (absDiff == 0 || absDiff > 3 || (diff > 0 !== increasing)) {
                safe = false;
                break;
            }
        }
    
        if (safe) {
            result++;
        }
    });

    console.log(result);
});

File: day2/part1.hs
module Main where
    
-- https://stackoverflow.com/questions/4503958/what-is-the-best-way-to-split-a-string-by-a-delimiter-functionally
splitBy :: (Foldable t, Eq a) => a -> t a -> [[a]]
splitBy delimiter = foldr f [[]] 
        where f c l@(x:xs) | c == delimiter = []:l
                           | otherwise = (c:x):xs

checkNumbers :: [Int] -> Int -> Int -> Int -> Bool
checkNumbers ns a b d = and [(abs (y - x) `elem` [a..b]) && signum (y - x) == d | (x, y) <- zip ns (tail ns)]

isSafe :: [Int] -> Bool
isSafe ns = checkNumbers ns 1 3 $ signum (ns !! 1 - ns !! 0)

main :: IO ()
main = do
    fileContents <- readFile "input.txt"
    let allNumbers = [[read x :: Int | x <- splitBy ' ' line] | line <- lines fileContents]
    print $ sum [1 | nums <- allNumbers, isSafe nums]